wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,304 @@
+            ["log", self.credentials.username.as_str()],
+            ["pwd", self.credentials.password.as_str()],
+            ["rememberme", "true"],
+            [
+                "redirect_to",
+                self.api_base_url.derived_rest_nonce_url().as_str(),
+            ],
+        ])
+        .map_err(|err| AuthenticationError::InvalidWebFormContent)?;
+
+        Ok(WpNetworkRequest {
+            method: RequestMethod::POST,
+            url: self.api_base_url.derived_wp_login_url().into(),
+            header_map: WpNetworkHeaderMap::new(headers).into(),
+            body: Some(WpNetworkRequestBody::new(body.into_bytes()).into()),
+        })
+    }
+
+    async fn nonce_from_request(&self, request: WpNetworkRequest) -> Option<String> {
+        self.request_executor
+            .execute(request.into())
+            .await
+            .ok()
+            .and_then(|response| {
+                // A 200 OK response from the `rest_nonce_url` (a.k.a `wp-admin/admin-ajax.php`)
+                // should be the nonce value. However, just in case the site is configured to
+                // return a 200 OK response with other content (for example redirection to
+                // other webpage), here we check the body length here for a light validation of
+                // the nonce value.
+                if response.status_code == 200 {
+                    let body = response.body_as_string();
+                    if body.len() < 50 {
+                        return Some(body);
+                    }
+                }
+                None
+            })
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for CookieAuthenticator {
+    async fn authenticate(
+        &self,
+        request: &WpNetworkRequest,
+    ) -> Result<HeaderMap, AuthenticationError> {
+        // Only attempt login if the request is to the WordPress site.
+        if let (Ok(api_base_url), Ok(request_url)) = (
+            url::Url::parse(self.api_base_url.as_str()),
+            url::Url::parse(request.url.0.as_str()),
+        ) {
+            if api_base_url.host_str() != request_url.host_str() {
+                return Err(AuthenticationError::RequestUrlDoesNotMatchSite);
+            }
+        }
+
+        self.get_rest_nonce().await.map(|nonce| {
+            let mut headers = HeaderMap::new();
+            headers.insert(
+                "X-WP-Nonce",
+                nonce
+                    .try_into()
+                    .expect("This conversion should never fail since nonce is a short string"),
+            );
+            headers
+        })
+    }
+
+    async fn re_authenticate(
+        &self,
+        request: &WpNetworkRequest,
+        previous_response: &WpNetworkResponse,
+    ) -> Result<HeaderMap, AuthenticationError> {
+        *self.nonce.write().expect("Failed to unlock nonce") = None;
+
+        return self.authenticate(request).await;
+    }
+}
+
+#[derive(Debug)]
+pub struct AuthenticatedRequestExecutor {
```

Comment:
> I wonder if you decided to introduce it due to my initial comment about the complicated logic in the WpDerivedRequest macro.

Yes. I like the current change because there are minimal changes to the macro implementation. To be honest, it's not very straightforward to make changes to `wp_derive_request_builder`.